### PR TITLE
Zero Scans migrate to new site.

### PR DESCRIPTION
**Pull request by FourTOne5**
Checklist:

- [X] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [ ] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [X] Have tested the modifications by compiling and running the extension through Android Studio

### CODE REVIEW

**Comment by FourTOne5** [comment]
The PR is ready for review.

File: src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+import okhttp3.Request
+import okhttp3.Response
+import rx.Observable
+import rx.Single
+import rx.schedulers.Schedulers
+import uy.kohesive.injekt.injectLazy
+
+class ZeroScans : HttpSource() {
+
+    override val name: String = "Zero Scans"
```
Comment (vetleledaal) [inline src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt:24]: the space changes the versionId IIRC

File: src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+            statusList = it.statuses.map { statusDto ->
+                Status(statusDto.name, statusDto.id)
+            }
+            comicList = it.comics
+            rankings = it.rankings
+        }
+    }
+
+    // Unused Stuff
+    override fun imageUrlParse(response: Response): String = throw Exception("Not Used")
```
Comment (vetleledaal) [inline src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt:306]: Maybe `throw UnsupportedOperationException("Not used.")` to make these explicit and conform with the contributing guidelines

**Review by vetleledaal — COMMENTED** [review]
Not sure about the null checks on Dto's -- mapNotNull for example doesn't seem necessary. Also, with how `getRankingsIfNeeded` is used type is never null.

File: src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt (hunk 1 above)
Comment (FourTOne5) [reply src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt:24]: It changes the id of the source, which I would've done by bumping the version id anyway.

File: src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+
+    override fun latestUpdatesParse(response: Response): MangasPage {
+        val newChapters = json.decodeFromString<NewChaptersResponseDto>(response.body!!.string())
+
+        return newChapters.all.mapNotNull { newChaptersManga ->
+            comicList.firstOrNull { comic ->
+                comic.slug == newChaptersManga.slug
+            }?.let { zsHelper.zsComicEntryToSManga(it) }
+        }.let { MangasPage(it, false) }
+    }
```
Comment (alessandrojean) [inline src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt:57]:
```suggestion
    override fun latestUpdatesParse(response: Response): MangasPage {
        val newChapters = json.decodeFromString<NewChaptersResponseDto>(response.body!!.string())

        val titlesList = newChapters
            .mapNotNull { comicList.firstOrNull { comic -> comic.slug == it.slug } }
            .map(zsHelper::zsComicEntryToSManga)
        
        return MangasPage(titlesList, hasNextPage = false)
    }
```

File: src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+            }
+        }
+
+        // Get 20 comics at a time
+        val chunkedFilteredComics = filteredComics.chunked(20)
+        if (chunkedFilteredComics.isEmpty()) {
+            return Observable.just(MangasPage(emptyList(), false))
+        }
+
+        val comics = chunkedFilteredComics[page - 1].map { zsHelper.zsComicEntryToSManga(it) }
```
Comment (alessandrojean) [inline src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt:109]:
```suggestion
        val comics = chunkedFilteredComics[page - 1].map(zsHelper::zsComicEntryToSManga)
```

File: src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+
+        if (!ascending) {
+            comics = comics.reversed()
+        }
+
+        return comics.mapNotNull { rankingEntry ->
+            comicList.firstOrNull { comic ->
+                comic.slug == rankingEntry.slug
+            }?.let { zsHelper.zsComicEntryToSManga(it) }
+        }
```
Comment (alessandrojean) [inline src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt:132]:
```suggestion
        return comics
            .mapNotNull { rankingEntry ->
                comicList.firstOrNull { comic -> comic.slug == rankingEntry.slug }
            }
            .map(zsHelper::zsComicEntryToSManga)
```

File: src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+        runCatching { updateComicsData() }
+        val mangaslug = "$baseUrl${manga.url}".toHttpUrl().pathSegments[1]
+
+        try {
+            comicList.first { it.slug == mangaslug }.let { zsComicEntry ->
+                zsHelper.zsComicEntryToSManga(zsComicEntry).let { return Observable.just(it) }
+            }
+        } catch (e: NoSuchElementException) {
+            throw Exception("Migrate from Zero Scans to Zero Scans")
+        }
```
Comment (alessandrojean) [inline src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt:145]:
```suggestion
        val mangaSlug = "$baseUrl${manga.url}".toHttpUrl().pathSegments[1]
        
        val manga = comicList.firstOrNull { it.slug == mangaSlug }
            ?.let { zsHelper.zsComicEntryToSManga(it) }
            ?: throw Exception("Migrate from Zero Scans to Zero Scans.")
            
        return Observable.just(manga)
```

File: src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+    override fun getFilterList(): FilterList {
+        return mutableListOf<Filter<*>>().apply {
+            if (genreList.isNotEmpty()) add(GenreFilter(genreList))
+            if (statusList.isNotEmpty()) add(StatusFilter(statusList))
+            add(SortFilter(sortList))
+            add(RankingsHeader())
+            add(RankingsHeader2())
+            add(RankingsFilter(rankingList))
+        }.let { FilterList(it) }
+    }
```
Comment (alessandrojean) [inline src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt:240]: It's better to focus on readability of the code.

```suggestion
    override fun getFilterList(): FilterList {
        val filters = mutableListOf()
        
        if (genreList.isNotEmpty()) {
            filters.add(GenreFilter(genreList))
        }
        
        if (statusList.isNotEmpty()) {
            filters.add(StatusFilter(statusList))
        }
        
        filters += listOf(
            SortFilter(sortList),
            RankingsHeader(),
            RankingsHeader2(),
            RankingsFilter(rankingList)
        )
        
        return FilterList(filters)
    }
```

File: src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt (hunk 2 above)
Comment (alessandrojean) [reply src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt:306]: Alternatively, could just return `""` too in this case. In the other ones, better to follow Vetle suggestion.

File: src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+        if (page == 1) runCatching { updateComicsData() }
+        return super.fetchLatestUpdates(page)
+    }
+
+    override fun latestUpdatesRequest(page: Int): Request {
+        return GET("$baseUrl/$API_PATH/new-chapters")
+    }
+
+    override fun latestUpdatesParse(response: Response): MangasPage {
+        val newChapters = json.decodeFromString<NewChaptersResponseDto>(response.body!!.string())
```
Comment (alessandrojean) [inline src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt:50]: I would suggest implementing the `Response.parseAs` custom method to reuse the code.

```kotlin
private inline fun <reified T> Response.parseAs(): T = use {
    json.decodeFromString(it.body?.string().orEmpty())
}
```

Then this one could be replaced by:

```kotlin
val newChapters = response.parseAs<NewChaptersResponseDto>()
```

**Review by alessandrojean — COMMENTED** [review]
Vast majority are just code readability suggestions, doesn't necessarily need to be followed.

File: src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+    override fun fetchMangaDetails(manga: SManga): Observable<SManga> {
+        runCatching { updateComicsData() }
+        val mangaSlug = "$baseUrl${manga.url}".toHttpUrl().pathSegments[1]
+
+        try {
+            val comic = comicList.first { comic -> comic.slug == mangaSlug }
+                .let { zsHelper.zsComicEntryToSManga(it) }
+
+            return Observable.just(comic)
+        } catch (e: NoSuchElementException) {
```
Comment (alessandrojean) [inline src/en/zeroscans/src/eu/kanade/tachiyomi/extension/en/zeroscans/ZeroScans.kt:149]: You can still remove this try-catch block by using `firstOrNull` instead. Rest seems fine.

**Pull request merged**